### PR TITLE
Testsuite: add separate definitions for menu steps

### PR DESCRIPTION
**Pull request by ncounter**
#### tl;dr
Standardize reaching the menu in the testsuite: migrate to one single way of following menu links, and separate menu steps in a proper file.

#### The long explanation
The shape of the menu is changed, and the behavior as well.
At the moment in the testsuite we have a *mixed flavor* ways of reaching menu links, and with the new changes some are failing and some other don't. To fix failures we can choose to fix case by case where they fail (and this will happen again ad again in the future as soon as something changes), or start using one and only one way for doing it.
This PR propose a single and dedicated approach for the testsuite menu steps.

Usage example: `When I follow the left menu "Configuration > Systems > Target"`

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: it's only about the testsuite

- [x] **DONE**

## Test coverage
- Cucumber tests were added: it is about the testsuite

- [x] **DONE**

## Links

Fixes #
Tracks # **add downstream PR, if any**

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"		 
- [ ] Re-run test "java_pgsql_tests"		 
- [ ] Re-run test "ruby_rubocop" 
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"		 
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"

### CODE REVIEW

File: testsuite/features/step_definitions/menu_steps.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Copyright (c) 2010-2019 Novell, Inc.
+# Licensed under the terms of the MIT license.
+
+#
+# Menu links
+#
+
+# Menu link: 1st level
+When(/^I follow the left menu "([^"]*)"$/) do |arg1, arg2|
```
Comment (srbarrios) [inline testsuite/features/step_definitions/menu_steps.rb]: arg2 is never used

Comment (Bischoff) [reply testsuite/features/step_definitions/menu_steps.rb]: and please don't use generic `argN`, but explicit names like `menuN`

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 2]
```diff
@@ -289,6 +289,44 @@
        and contains(@class,'node')]/ul/li/div/a[contains(.,'System List')]").click
 end
 
+#
+# Menu links
+#
+
+# Menu link: 1st level
+When(/^I follow the left menu "([^"]*)"$/) do |arg1|
```
Comment (srbarrios) [inline testsuite/features/step_definitions/navigation_steps.rb]: It would be nice to reduce it in only one method, parse the argument and split using ">"

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 3]
```diff
@@ -289,6 +289,44 @@
...
 end
 
+#
+# Menu links
+#
+
+# Menu link: 1st level
+When(/^I follow the left menu "([^"]*)"$/) do |menu_level1|
+  xpath_menu_level1 = "//aside/div[@id='nav']/nav/ul/li/div/a[contains(.,'#{menu_level1}')]" # reach the link
+  find(:xpath, xpath_menu_level1).click
```
Comment (srbarrios) [inline testsuite/features/step_definitions/navigation_steps.rb]: "find" method will be critical, it will be call tons of times.
We should be sure we have it optimised, probably is just fine like that, but I think "find" inspect the whole page, maybe better to use "first". Something to test and compare ;)

File: testsuite/features/step_definitions/navigation_steps.rb (hunk 2 above)
Comment (Bischoff) [reply testsuite/features/step_definitions/navigation_steps.rb]: Do we have more than 3 ?

File: testsuite/features/step_definitions/navigation_steps.rb (hunk 3 above)
Comment (Bischoff) [reply testsuite/features/step_definitions/navigation_steps.rb]: Right.

File: testsuite/features/step_definitions/navigation_steps.rb (hunk 2 above)
Comment (ncounter) [reply testsuite/features/step_definitions/navigation_steps.rb]: No no, luckily we don't, and I would fight against having more :wink:

File: testsuite/features/step_definitions/menu_steps.rb (hunk 1 above)
Comment (ncounter) [reply testsuite/features/step_definitions/menu_steps.rb]: Done. Good catch, thanks.

File: testsuite/features/step_definitions/navigation_steps.rb (hunk 2 above)
Comment (srbarrios) [reply testsuite/features/step_definitions/navigation_steps.rb]: Having the same code repeated in three methods is not clean code, it's harder to maintain.

Comment (ncounter) [reply testsuite/features/step_definitions/navigation_steps.rb]: I totally agree @srbarrios .
This was the initial proposal to **firstly** unify the way we search for menu links and how it should behave. Then I see improvements and refactoring as `tech-debt`, but they are also on their way (don't hold your breath btw :wink )

Comment (Bischoff) [reply testsuite/features/step_definitions/navigation_steps.rb]: Factorizing to the excess is not always a good idea...

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 4]
```diff
@@ -283,12 +281,55 @@
   page.execute_script("ace.edit('contents-editor').insert('#{arg1}')")
 end
 
-When(/^I click System List, under Systems node$/) do
-  find(:xpath, "//div[@id='nav']/nav/ul/li[contains(@class, 'active')
-       and contains(@class, 'open')
-       and contains(@class,'node')]/ul/li/div/a[contains(.,'System List')]").click
+#
+# Menu links
```
Comment (srbarrios) [inline testsuite/features/step_definitions/navigation_steps.rb]: I didn't tested, but I would simply the code with something like that:

´´´
When(/^I follow "([^"]*)" in the left menu$/) do |menu_path|
  menu_steps = menu_path.split('>').map(&:strip)
  menu_steps.each do |step|
    within(:xpath, '//aside') do
      step_parent = "#{step}/parent::div/parent::li"
      unless all(:xpath, step_parent).any? and find(:xpath, step_parent)[:class].include?('open')
        click_link(step, wait: CLICK_TIMEOUT, match: :first)
      end
    end
  end
end
´´´

File: testsuite/features/step_definitions/navigation_steps.rb (hunk 2 above)
Comment (srbarrios) [reply testsuite/features/step_definitions/navigation_steps.rb]: Oki, cool, just take a look tomorrow to my suggested refactor for those three methods ;)

Comment (ncounter) [reply testsuite/features/step_definitions/navigation_steps.rb]: @srbarrios Absolutely yes. The direct link is still there, it is just squashed with the menu element text.

The new menu behaves like this:

```html
<ul>
  <li>
    <div> <!-- click here to open/close the submenu, if any -->
      <a>Level1</a> <!-- click here to load directly the first element of the submenu, Level2-1 in this case -->
    </div>
    <ul>
      <li>
        <div>
          <a>Level2-1</a>
        </div>
      </li>
      <li>
        <div>
          <a>Level2-2</a>
        </div>
      </li>
    </ul>
  </li>
</ul>
```

This means that if the step says
`When I follow the left menu "Level1" > "Level2-1"`
or just
`When I follow the left menu "Level1"`
that's the same thing in the end, because of the new implementation where if you pass the second level argument it will open the submenu and simulate the click on the sublevel, if you don't it will just trigger the click on the first level link.

The problem is different if you want to load a different page than the first option in the submenu, let's say `Level2-2`, then you will have to give the complete path and in that case I think it is still useful to do the `open/close` visibility of the submenu.

Bottom line: for the testsuite usage, I agree it will be faster to use the direct link click. Let's do the following: right now I refactored inserting everywhere (where I did not overlooked :wink: ) the complete menu path. Later we can still remove the last argument where we want to use the direct link instead. What do you think?

File: testsuite/features/core_srv_menu.feature [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2017-2019 SUSE LLC
+# Licensed under the terms of the MIT license.
+
+Feature: Web UI menu
```
Comment (Bischoff) [inline testsuite/features/core_srv_menu.feature]: 1. I think this should go to secondary features - it's not a preparation
2. We already have a feature that does check menus - please merge both features (as secondary)

File: testsuite/features/core_srv_menu.feature [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    Then I should see a "Centrally-Managed Configuration Files" text in the content area
+    And I should not see a "Locally Managed Configuration Files" text in the content area
+
+  Scenario: Test idempotency of complete menu path and direct link
+    Given I am authorized as "admin" with password "admin"
+    When I follow the left menu "Admin" > "Manager Configuration"
+    Then I should see a "SUSE Manager Configuration - General Configuration" text in the content area
+    When I follow the left menu "Admin" > "Manager Configuration" > "General"
+    Then I should see a "SUSE Manager Configuration - General Configuration" text in the content area
+
```
Comment (Bischoff) [inline testsuite/features/core_srv_menu.feature]: No final linefeed please

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 7]
```diff
@@ -16,6 +16,12 @@
   end
 end
 
+When(/^I should not see a "(.*)" text in the content area$/) do |txt|
```
Comment (Bischoff) [inline testsuite/features/step_definitions/navigation_steps.rb]: Then

File: testsuite/features/step_definitions/common_steps.rb [hunk 8]
```diff
@@ -373,7 +373,7 @@
 
 Then(/^I should see package "([^"]*)" in channel "([^"]*)"$/) do |pkg, channel|
   steps %(
-    And I follow "Channel List > All" in the left menu
+    And I follow the left menu "Software" > "Channel List" > "All"
```
Comment (Bischoff) [inline testsuite/features/step_definitions/common_steps.rb]: When

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 9]
```diff
@@ -354,16 +399,14 @@
 Given(/^I am on the groups page$/) do
   steps %(
     Given I am on the Systems page
-    And I follow "System Groups" in the left menu
+    And I follow the left menu "Systems" > "System Groups"
```
Comment (Bischoff) [inline testsuite/features/step_definitions/navigation_steps.rb]: When

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 10]
```diff
@@ -354,16 +399,14 @@
...
     )
 end
 
 Given(/^I am on the active Users page$/) do
   steps %(
     Given I am authorized as "admin" with password "admin"
-    And I follow "Users"
-    And I follow "User List"
-    And I follow "Active"
+    And I follow the left menu "Users" > "User List" > "Active"
```
Comment (Bischoff) [inline testsuite/features/step_definitions/navigation_steps.rb]: When

File: testsuite/features/trad_baremetal_discovery.feature [hunk 11]
```diff
@@ -14,9 +14,7 @@ Feature: Bare metal discovery
 
   Scenario: Enable bare metal discovery
     Given I am authorized as "admin" with password "admin"
-    And I follow "Admin"
-    And I follow "Manager Configuration" in the left menu
-    And I follow "General" in the left menu
+    And I follow the left menu "Admin" > "Manager Configuration" > "General"
```
Comment (Bischoff) [inline testsuite/features/trad_baremetal_discovery.feature]: When

File: testsuite/features/trad_baremetal_discovery.feature [hunk 12]
```diff
@@ -36,8 +34,7 @@ Feature: Bare metal discovery
 
   Scenario: See the client in unprovisioned systems list
     Given I am on the Systems page
-    And I click System List, under Systems node 
-    And I follow "Unprovisioned Systems" in the left menu
+    And I follow the left menu "Systems" > "System List" > "Unprovisioned Systems"
```
Comment (Bischoff) [inline testsuite/features/trad_baremetal_discovery.feature]: When

File: testsuite/features/trad_baremetal_discovery.feature [hunk 13]
```diff
@@ -53,7 +50,7 @@ Feature: Bare metal discovery
 
   Scenario: Check unprovisioned system details
     Given I am on the Systems page
-    And I click System List, under Systems node 
+    And I follow the left menu "Systems" > "System List"
```
Comment (Bischoff) [inline testsuite/features/trad_baremetal_discovery.feature]: When

File: testsuite/features/trad_baremetal_discovery.feature [hunk 14]
```diff
@@ -111,9 +108,7 @@ Feature: Bare metal discovery
 
   Scenario: Cleanup: disable bare metal discovery
     Given I am authorized as "admin" with password "admin"
-    And I follow "Admin"
-    And I follow "Manager Configuration" in the left menu
-    And I follow "General" in the left menu
+    And I follow the left menu "Admin" > "Manager Configuration" > "General"
```
Comment (Bischoff) [inline testsuite/features/trad_baremetal_discovery.feature]: When

File: testsuite/features/trad_migrate_to_sshminion.feature [hunk 15]
```diff
@@ -20,8 +20,7 @@ Feature: Migrate a traditional client into a Salt SSH minion
 
   Scenario: Change contact method of activation key to ssh-push
     Given I am authorized as "admin" with password "admin"
-    And I follow "Systems" in the left menu
-    And I follow "Activation Keys"
+    And I follow the left menu "Systems" > "Activation Keys"
```
Comment (Bischoff) [inline testsuite/features/trad_migrate_to_sshminion.feature]: When

File: testsuite/features/trad_ssh_push.feature [hunk 16]
```diff
@@ -54,15 +54,15 @@ Feature: Register a traditional system to be managed via SSH push
 
   Scenario: Cleanup: delete the activation key for SSH push
     Given I am on the Systems page
-    And I follow "Activation Keys" in the left menu
+    And I follow the left menu "Systems" > "Activation Keys"
```
Comment (Bischoff) [inline testsuite/features/trad_ssh_push.feature]: When

File: testsuite/features/trad_ssh_push.feature [hunk 17]
```diff
@@ -54,15 +54,15 @@ Feature: Register a traditional system to be managed via SSH push
...
     And I follow "SSH push key" in the content area
     And I follow "Delete Key"
     And I click on "Delete Activation Key"
     And I should see a "Activation key SSH push key has been deleted." text
 
   Scenario: Cleanup: delete the activation key for SSH push via tunnel
     Given I am on the Systems page
-    And I follow "Activation Keys" in the left menu
+    And I follow the left menu "Systems" > "Activation Keys"
```
Comment (Bischoff) [inline testsuite/features/trad_ssh_push.feature]: When

File: testsuite/run_sets/testsuite.yml [hunk 18]
```diff
@@ -10,6 +10,7 @@
 
 - features/core_first_settings.feature
 # initialize SUSE Manager server
+- features/core_srv_menu.feature
```
Comment (Bischoff) [inline testsuite/run_sets/testsuite.yml]: To be made secondary, and merged with the other feature that does also check menus.

File: testsuite/features/step_definitions/navigation_steps.rb (hunk 4 above)
Comment (ncounter) [reply testsuite/features/step_definitions/navigation_steps.rb]: @srbarrios I refactored the steps into one dynamic only. I reuse some of your logic but still I need to keep the `find` and `click` on parent when the submenu needs to be open. You can take a look [here](https://github.com/uyuni-project/uyuni/pull/937/commits/24b0b63404054089ab499792fea0328592e01e7c#diff-69eaaf5982104837a8e6521c7fa3d4feR295)

File: testsuite/features/core_srv_menu.feature (hunk 5 above)
Comment (ncounter) [reply testsuite/features/core_srv_menu.feature]: >     1. I think this should go to secondary features - it's not a preparation

I had the same thinking, but then: why not to test it at the very begin? if the menu is broken, the most of the tests will fail as well. I though about this: I start the testsuite run, I see the menu is broken, I quickly abort the running and fix it soon. If you think this is not that much relevant, I will just move it to a secondary feature, as I agree about it's not a preparation.

Comment (srbarrios) [reply testsuite/features/core_srv_menu.feature]: For now fine to keep at core section yep.
But when we decide to use the testsuite in our PRs, then the developer will want to run the tests more important for their changes excluding whatever it takes unnecessary extra time. We are not at that point yet, so fine to let it at core.

File: testsuite/run_sets/testsuite.yml (hunk 18 above)
Comment (ncounter) [reply testsuite/run_sets/testsuite.yml]: Addressed in https://github.com/uyuni-project/uyuni/pull/937/commits/dad89c163b875c03668ca18b273e96321ad88792

File: testsuite/features/core_srv_menu.feature (hunk 5 above)
Comment (ncounter) [reply testsuite/features/core_srv_menu.feature]: Nice POV, I didn't thought about it. Great, I just squashed all menu scenarios into one in https://github.com/uyuni-project/uyuni/pull/937/commits/dad89c163b875c03668ca18b273e96321ad88792 . We will move it down in the order once testsuite on PRs will be in place :+1:

Comment (Bischoff) [reply testsuite/features/core_srv_menu.feature]: I suggest a compromise. Why not move it very top of secondary ?

That way, if a developer just wants a setup, he/she will not lose the time testing the menus. But it would still be tested very early.

Is that acceptable?

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 19]
```diff
@@ -308,15 +308,16 @@
...
-    # if this is not the last element of the path
-    unless index == (menu_levels.count - 1)
-      # open the submenu if needed
-      unless find(:xpath, target_link_path + parent_wrapper_path + parent_level_path)[:class].include?('open')
-        find(:xpath, target_link_path + parent_wrapper_path).click
-      end
-      # point the target to the current menu level
-      target_link_path += parent_wrapper_path + parent_level_path
+    # if this is the last element of the path
+    if index == (menu_levels.count - 1)
```
Comment (srbarrios) [inline testsuite/features/step_definitions/navigation_steps.rb]:
```
break if index == (menu_levels.count - 1)
```
